### DISCUSSION TITLE
Required audio format for Deepgram real-time transcription

### DISCUSSION THREAD

**Discussion by ductuannguyen1999** [deepgram/community#233] [General help]
Hi Team,

I'm building a chrome extension that records the browser's audio using the MediaRecorder interface. Then in the "dataavailable" event, I send the audio data to one of my services for transcription purpose. Since I cannot send the raw data from the event retrieved from the media recorder, I converted the raw data into binary string, send to the destination service, and convert the binary string back to Uint8Array binary data. I then send this binary data to Deepgram's real-time transcription socket, but sometimes (not always) the socket closes unexpectedly with the response payload like below:   

`{"transaction_key":"deprecated","request_id":"3aa0b400-5520-45cc-b084-bbcf74006acc","sha256":"d9364c58d43e33499cc01550fd4a68fd30af8f94efb1703b30032a3cc47618b5","created":"2023-06-26T14:52:29.327Z","duration":0.0,"channels":0}  `

Does Deepgram real-time transcription socket allow Uint8Array like what I described above? If it does, then is there something wrong occured in my process of converting the audio data? Could anyone help me with this? I would really appreciate your help!

**Comment by lukeocodes** [comment]
Off the top of my head, check this out: https://github.com/deepgram-devs/node-live-example It uses MediaRecorder and sends Deepgram live-transcription the data. We don't have a client SDK right now, but you may be able to make a direct-integration (or use a small server like this example does).

**Reply by ductuannguyen1999** [reply]
Thank you for the repository! I will definitely look at it. However, I would like to understand the root cause of this issue. Do you have any idea why the response is returning "transaction_key" being "deprecated" ? Is there something wrong in my process described in the question that is causing this issue ?

**Reply by nikolawhallon** [reply]
The "transaction_key" field is a red herring here - it is a field which isn't used anymore, but hasn't been removed in order to avoid a breaking API change, so it will always come back as "deprecated" - the fact that the response indicates a duration of 0.0 and 0 channels likely means Deepgram was not able to understand the data so immediately closed the connection, sending back this metadata response.

**Comment by sorokinvj** [comment]
we are doing the same thing, and we stuck with the same problem. Sometime Deepgram just unexpectedly closes the connection. 
btw we do our websocket server which basically forwards all requests to Deepgram. However, we are able to do some extra, for example I am trying to analyze audio chunks from the client app with relation to audio headers:

```
  socket.on('message', async (message: Message) => {
    if (message.type === MessageType.AUDIO) {
      try {
        // get info about audio - codec, sample rate, bit depth
        audioAnalyzer.analyzeAudioChunk(message.data);

        // send audio to transcription service
        deepgramService.transcribeAudio(message.data);
      } catch (error) {
        logger.error(error);
      }
    }
  });
```

And I noticed that sometimes there are just no headers! and this is exactly the time when Deepgram is stuck and don't return transcriptions. I understand that missing headers could be somewhat the reason, especially when Deepgram docs are saying that these headers are important - https://developers.deepgram.com/docs/determining-your-audio-format-for-live-streaming-audio

so I thought, ok, I cannot control MediaRecorder and check that every audio chunk is having header, well maybe I can do it by myself, and I started to send raw PCM16 data with a special WAV header on every request. But then hit the wall, because the latency of transcriptions skyrocketed to 7 seconds! with webm/opus it was equal to manageable 4s (though also quite 4 seconds, when you need a real time thing). 

so my question now, and maybe @lukeocodes you could help with that: do you guys know what combination of codec/sample_rate/bit_rate produces the fastest result? How could you explain the increase of the latency from 4s (`web/opus`) to 7s(`wav, 16pcm`)?

here is my server btw: https://github.com/sorokinvj/transcription-server

**Reply by jjmaldonis** [reply]
Hey @sorokinvj, can you send us the audio so we can reproduce this? It may have something to do with how much data is expected per unit time for the different audio types. To dig in, it would be very helpful to have a standalone reproduceable example.
